nsot/models.py

```diff
@@ -1718,6 +1719,186 @@ def to_dict(self):
+    sender=ProtocolType.required_attributes.through
+)
+
+
+class Protocol(Resource):
+    """
+    Represetation of a routing protocol running over a circuit
+    """
+
+    site = models.ForeignKey(
+        Site, db_index=True, related_name='protocols',
+        on_delete=models.PROTECT,
+        help_text='Unique ID of the Site this Protocol is under.',
+    )
+    type = models.ForeignKey(
+        ProtocolType, db_index=True, related_name='protocols',
+    )
+    device = models.ForeignKey(
+        Device, db_index=True, null=False, related_name='protocols',
+        help_text='Device that this protocol is running on',
+    )
+    interface = models.ForeignKey(
+        Interface, db_index=True, blank=True, null=True,
+        related_name='protocols',
+        help_text=(
+            'Interface this protocol is running on. Either interface or'
+            ' circuit must be populated.'
+        ),
+    )
+    circuit = models.ForeignKey(
+        Circuit, db_index=True, blank=True, null=True,
+        related_name='protocols',
+        help_text='Circuit that this protocol is running over.',
+    )
+    auth_string = models.CharField(
+        max_length=255, default='', blank=True,
+        help_text='Authentication string (such as MD5 sum)',
+    )
+    description = models.CharField(
+        max_length=255, default='', blank=True,
+        help_text='Description for this Protocol'
+    )
+
+    def __unicode__(self):
+        return u'%s over %s' % (self.get_type_display(), self.circuit)
+
+    class Meta:
+        ordering = ('device', )
+
+    def local_interface(self):
+        """
+        Returns the local interface for this Protocol, either the set interface
+        or the local interface of circuit
+        """
+        if self.interface:
+            return self.interface
+        elif self.circuit is None:
+            return None
+
+        for endpoint in (self.circuit.endpoint_a, self.circuit.endpoint_z):
+            if endpoint.device.id == self.device.id:
+                return endpoint
+
+        return None
+
+    def remote_interface(self):
+        """
+        Returns the remote device's interface attached to the circuit
+        """
+        if self.circuit is None:
+            return None
+
+        for endpoint in (self.circuit.endpoint_a, self.circuit.endpoint_z):
+            if endpoint.device.id != self.device.id:
+                return endpoint
+
+        return None
+
+    def clean_site(self, value):
+        if value is None:
+            return self.device.site_id
+
+        return value
+
+    def clean_circuit(self, value):
+        """ Ensure at least one endpoint on the circuit is on this device """
+        if value and self.local_interface() is None:
+            raise exc.ValidationError({
+                'circuit': (
+                    'At least one endpoint of the circuit must match the '
+                    'device'
+                )
+            })
+
```

Comment:
Wouldn't this check pass if OSPF/ISIS was the protocol type and a `circuit` with one endpoint on the box was specified?

Reply:
in this case `clean_circuit()` is a bit of a misnomer since it's using `local_interface()` that currently returns either the set `interface` or the local interface on the `circuit`. This as-is is something I didn't update after adding `interface` so I need to re-work this. 

But in regards to this and your comment on `clean_interface()`, they _should_ pass in the cases you specify. I'm trying to be unopinionated with this Protocol object to allow users to use it how it fits in with their use-cases.

If you feel strongly about this sort of validation, we could possibly add something to `ProtocolType` to allow users to do validation on concrete `Protocol` fields in addition to attributes.